www/adblock2privoxy/Portfile

```diff
@@ -39,18 +40,54 @@ depends_run-append \
                     port:nginx \
                     port:privoxy
 
+variant initialize \
+    description {Initialize all configuration files. Existing
+        configurations files are not overwritten by default.} {}
+
+# destroot approach without destroot capability in cabal (or stack):
+# create symlink in ${prefix}, install using symlink, delete symlink;
+# if the build fails before these links are deleted, delete them by hand (below)
+extract.asroot      yes
+
+# Without `--datadir`, the adblock2privoxy binary hardcodes file locations
+# ab2p.system.action and .task within the temporary haskell_stack.stack_root
+post-extract {
+    xinstall -o macports -m 0755 -d \
+        "[option haskell_stack.stack_root]" \
+        ${destroot}${prefix}/share/${name}
+
+    # symbolic link for destroot
+    delete ${prefix}/share/${name}
```

Comment:
This is very dangerous. You cannot do this. Imagine that somebody has an older version of this port installed and runs `sudo port extract`. Suddenly, all files installed by this port in this directory are gone, and there is no simple way to get them back.

Reply:
@neverpanic 
> This is very dangerous. You cannot do this. Imagine that somebody has an older version of this port installed and runs `sudo port extract`. Suddenly, all files installed by this port in this directory are gone, and there is no simple way to get them back.

The only files installed in this port-specific share directory are the ones that `stack build` creates, and the simple way to get them back is just to run `port install':
```
port contents adblock2privoxy | grep /opt/local/share/adblock2privoxy
  /opt/local/share/adblock2privoxy/x86_64-osx-ghc-8.6.5/adblock2privoxy-2.0.1/templates/ab2p.system.action
  /opt/local/share/adblock2privoxy/x86_64-osx-ghc-8.6.5/adblock2privoxy-2.0.1/templates/ab2p.system.filter
```

 I don't see any specific issue here.

I only included the `delete` step for robustness, and did this in the `extract` phase to limit the phases run as root. I could:
* Omit the `delete` line (what will happen during `port update`? What happens if you run `port extract` when the port is already installed?)
* Move to a `.previous` then move it back, rather than delete

We need some consensus about how to hack in a destrooted `--datadir` within Haskell world ports. Some hack compromise that manipulates specific links or directories within `${prefix}` during extract or configure or build or destroot will be necessary.

The benefits of the approach in the current commit:
* Only manipulate `${prefix}` within the extract phase as root
* Accomplishes `stack` destroot correctly with a `--datadir`
* Accomplishes updates robustly
* Nondestructive, except for files that are installed by the port anyway

Given the tradeoffs of `delete` in the extract phase and the limited impact, do you still believe that this is unacceptably dangerous? If so, are you okay with the issues that omitting the `delete` step will raise (error on `port extract` if the port is already installed)?

Reply:
@neverpanic I omitted the delete, added a conditional to handle `port extract` if the port is already installed.

Reply:
> The only files installed in this port-specific share directory are the ones that `stack build` creates, and the simple way to get them back is just to run `port install':
> 
> ```
> port contents adblock2privoxy | grep /opt/local/share/adblock2privoxy
>   /opt/local/share/adblock2privoxy/x86_64-osx-ghc-8.6.5/adblock2privoxy-2.0.1/templates/ab2p.system.action
>   /opt/local/share/adblock2privoxy/x86_64-osx-ghc-8.6.5/adblock2privoxy-2.0.1/templates/ab2p.system.filter
> ```

No, see, we're not talking about the same thing here. Let's assume you're going to update this port. A user has the port installed already and runs `sudo port upgrade outdated`. This port starts building, which deletes `/opt/local/share/adblock2privoxy`. For some reason the build gets interrupted (e.g. because the machine crashes, the build fails, the network connectivity is down, the user cancels the installation, or the user didn't run `sudo port install` in the first place, but only `sudo port destroot`).

The user's assumption will be because the build didn't finish, the old port is still there and untouched, but in fact your Portfile deleted the datadir, which now broke an already installed port. Even if the user notices what's broken, there's no simple way to get back to a sane state, because if you run `sudo port install adblock2privoxy` in this state, MacPorts will just tell you it is already installed. The only way to get out of this situation is to run `sudo port -f deactivate adblock2privoxy && sudo port activate adblock2privoxy`.

This is an unacceptable hack for me.
Additionally, it may not even work, because we're sandboxing our build and destroot phases to being unable to write anywhere but `${destroot}`.

> We need some consensus about how to hack in a destrooted `--datadir` within Haskell world ports. Some hack compromise that manipulates specific links or directories within `${prefix}` during extract or configure or build or destroot will be necessary.

And I'm telling you, I'm not going to merge anything that modifies `${prefix}` during a build. You'll have to find a different solution, either by beating stack until it does what it needs to do, or by using some other mechanism (bind mounts, union mounts, DYLD_INSERT_LIBRARIES, chrooting, ...).

> Given the tradeoffs of `delete` in the extract phase and the limited impact, do you still believe that this is unacceptably dangerous? If so, are you okay with the issues that omitting the `delete` step will raise (error on `port extract` if the port is already installed)?

We have the `conflicts_build` PortGroup to denote such conflicts, and there is precedent (poppler) for `conflicts_build ${name}`, which I guess could be used in this situation. However, that is annoying for users, and it still doesn't change that even without deleting files, you should still not create files in `${prefix}` from within a build phase, either. The only phase that should be touching `${prefix}` is activation.

Reply:
@neverpanic 

Note that the `delete` has been deleted now, which will prevent the worst outcomes you describe.

Would you please go back and take a look? If you still have concerns about the symbolic link created in `extract`, this will approximate my response:

I think this is too strict. The Portfile has been modified so that all it does during the extract phase is install a symbolic link, then deletes that same link when it's done.

Yes, it's a hack because `stack` doesn't do destroot. I honestly don't see how any of the possible solutions you posted would work. Look in the current binary for this port. There's a hardcoded directory to the original `${worksrcpath}` stuck in there, and a call to `install_name_tool` or building within chroot or any of those other options isn't going to get rid of it. They've been banging on about this over at stack and elsewhere for over five years now, and if there's a clever solution, no one has posted it yet. I sure don't have one. Someone even posted some method of overwriting the directory name (in the binary!) with the correct `PREFIX`-ed string followed by the right number of `NULL`s so as not to change the field length (!!) Believe me, there are far, far worse hacks out there…

Practically, such a restriction now will prevent several useful ports like this one, ihaskell, and others from adoption into MacPorts. And this one should be withdrawn because it's currently broken.

I suggest thinking about necessary flexibility until Haskell world adopts `DESTDIR` and `PREFIX`.

The suggested solution until they do is to create a symbolic link if the port isn't installed. I don't think that would cause the issues you raise now that that `delete` line is gone.

Reply:
@neverpanic This would leave `hlint` out too. See this related open issue: https://github.com/ndmitchell/hlint/issues/699

Reply:
@neverpanic I just pushed a new approach that I believe addresses all your comments: nothing destructive, everything isolated within the destroot phase, all working and robust I believe. Please take a look.

Also, I’ll ping the relevant open issues at https://github.com/haskell/cabal and https://github.com/commercialhaskell/stack about this ongoing problem. The earliest of these were opened in 2012 so even though they recognize the deficiency, there appears to be no one working it and I wouldn’t hold my breath for the fix.
* https://github.com/haskell/cabal/issues/462
* https://github.com/haskell/cabal/issues/3586
* https://github.com/commercialhaskell/stack/issues/848
* https://github.com/commercialhaskell/stack/issues/4857

Reply:
@neverpanic Further, I added a default port variant that explicitly blocks off the cabal destroot fix, and used portgroup `conflicts_build` as in port `poppler`.

When Haskell world provides destroot support, we can simply remove this variant and any Haskell ports built with `stack` or `cabal` will build normally.

Reply:
@neverpanic I’ve received a couple responses to my posts to Haskell world about this. This comment I believe represents t the current and expected future status, https://github.com/haskell/cabal/issues/462#issuecomment-527866197:
> No one is actively working on this AFAIK, but I think it'd make a good GSoC project.

Read: no work or plans to work on it. Maybe something after summer 2020.

Draw your own conclusions.

In the meantime, we have a fix here that works and I believe addresses your concerns above.

Would you please take a look?

I’d like to merge this port both to fix its current brokenness, use it within a larger proxy project I have in mind, and provide a path to fix Haskell ports that have broken destroot.

Reply:
@neverpanic FWIW homebrew currently addresses this `stack` issue by actually installing in `prefix`. See https://github.com/Homebrew/homebrew-core/pull/1630#issuecomment-223244572. This approach doesn’t do anything so drastic.

Reply:
@neverpanic 
> I'm telling you, I'm not going to merge anything that modifies `${prefix}` during a build.

Okay, thanks for being a hardass about this. 

I found and adopted an old [post](http://neilmitchell.blogspot.com/2008/02/adding-data-files-using-cabal.html) from @ndmitchell ([hlint](https://github.com/ndmitchell/hlint)'s author) that solves the problem.

I don't understand why this couldn't have been adopted into either `stack` or `cabal` years ago…

All that's required is to specify the path within a file called `Paths_packagename.hs`, which is normally automatically generated by `cabal` with its own paths. (Another approach that works is to set the environment variable `packagename_datadir` at runtime, which overrides the binary's hardcoded path, but we don't need this approach here.)

For posterity, setting `datadir` in `stack.yaml` causes `stack to actually try to write to that directory during compilation, which was the source of all the hack backflips attempted in previous commits above. All that's unnecessary with the `Paths_packagename.hs` file. I'll file an issue over at `stack`.

Furthermore, this approach will work in general, and I believe that it can be used in other ports that specify `data-files` in their `package name.cabal` file. Refactoring the `ihaskell` submission` will be the next test.

Ultimately, this is something that I think `stack` should do on-the-fly.

But this port is now fixed, and without any ugly hacks.